src/category_theory/monoidal/of_has_finite_products.lean

```diff
@@ -16,48 +16,92 @@ A category with a monoidal structure provided in this way is sometimes called a
 although this is also sometimes used to mean a finitely complete category.
 (See <https://ncatlab.org/nlab/show/cartesian+category>.)
 
-As this works with either products or coproducts, we don't set up either construct as an instance.
--/
+As this works with either products or coproducts,
+and sometimes we want to think of a different monoidal structure entirely,
+we don't set up either construct as an instance.
 
-open category_theory.limits
+## Implementation
+For the sake of nicer definitional properties,
+we rely on `has_terminal` and `has_binary_products` instead of `has_finite_products`,
+so that if a particular category provides customised instances of these
+we pick those up instead.
+-/
 
 universes v u
 
 namespace category_theory
+open category_theory.limits
 
-section
 variables (C : Type u) [𝒞 : category.{v} C]
 include 𝒞
 
+section
 local attribute [tidy] tactic.case_bash
 
-/-- A category with finite products has a natural monoidal structure. -/
-def monoidal_of_has_finite_products [has_finite_products.{v} C] : monoidal_category C :=
-{ tensor_unit  := terminal C,
-  tensor_obj   := λ X Y, limits.prod X Y,
+/-- A category with a terminal object and binary products has a natural monoidal structure. -/
+def monoidal_of_has_finite_products [has_terminal.{v} C] [has_binary_products.{v} C] : monoidal_category C :=
+{ tensor_unit  := ⊤_ C,
+  tensor_obj   := λ X Y, X ⨯ Y,
   tensor_hom   := λ _ _ _ _ f g, limits.prod.map f g,
   associator   := prod.associator,
   left_unitor  := prod.left_unitor,
   right_unitor := prod.right_unitor,
   pentagon'    := prod.pentagon,
   triangle'    := prod.triangle,
   associator_naturality' := @prod.associator_naturality _ _ _, }
+end
+
+namespace monoidal_of_has_finite_products
+variables [has_terminal.{v} C] [has_binary_products.{v} C]
+local attribute [instance] monoidal_of_has_finite_products
+
+@[simp]
+lemma left_unitor_hom (X : C) : (λ_ X).hom = limits.prod.snd := rfl
+@[simp]
+lemma right_unitor_hom (X : C) : (ρ_ X).hom = limits.prod.fst := rfl
+@[simp]
+lemma associator_hom (X Y Z : C) :
```

Comment:
Should this really be a simp-lemma? The right hand side looks pretty complicated. I'm not sure if I would like to have that end up in my goal state.

Reply:
It's true, but the point is that you hope that a whole lot of other simp lemmas then take over on that right hand side...

In concrete categories this should certainly happen. But I guess it's safest if we just define those sort of lemmas one by one, and mark them as simp.

Reply:
I've removed `@[simp]` for now.